### PR TITLE
Datastore entry method

### PR DESCRIPTION
**Pull request by feipenghe**
This PR fixes https://github.com/asyml/forte/issues/767. 

### Description of changes
* implemented method `all_entries(entry_type_name)`
* implemented method `num_entries(entry_type_name)`

### Possible influences of this PR.
Describe what are the possible side-effects of the code change.

### Test Conducted
* test `all_entries(entry_type_name)` returns an iterator of raw entry data in the `List` format of `entry_type_name`
* test `num_entries(entry_type_name)` returns a correct number of entries of `entry_type_name`

### CODE REVIEW

**Review by mylibrar — COMMENTED** [review]
This might not preserve the previous behavior in `DataPack` since it does not retrieve the subclasses. For example, `DataPack.all_annotations` is expected to return an iterator of all entries that satisfy `issubclass(Annotation)` (e.g., `Token`, `Document`, `Sentence`, etc.). Hence `DataStore.all_entries("forte.data.ontology.top.Annotation")` should also be able to get all subtypes of `Annotation` in its outputs.

File: tests/forte/data/data_store_test.py
Comment (mylibrar) [inline tests/forte/data/data_store_test.py:343]: This might not be correct if there are `Link`s or `Group`s in `DataStore`. For example, the format of  test entry with `tid=10123` looks like a `Group` but will also be included in `ann_list`

Comment (feipenghe) [reply tests/forte/data/data_store_test.py:343]: Is there any way to check `Link`, I found `Link` type entry in `self.__elements` is similar to a regular entry.

Comment (mylibrar) [reply tests/forte/data/data_store_test.py:343]: According to the docstring in `DataStore.__init__`, a `Link` entry has the following format: `[<parent_tid>, <child_tid>, <tid>, <type_name>, <attr_1>, <attr_2>, ..., <attr_n>, index_id]`. So I guess you can still use `<type_name>` to check if it's a link.

File: forte/data/data_store.py
Comment (hunterhector) [inline forte/data/data_store.py:503]: Can we create a `get_all_subclass` function here?

1. It will be used in multiple places
2. We can create a more efficient implementation later (you can leave the implementation now)
3. I think we need to sort the entry keys, otherwise the iteration order is kinda random

File: forte/data/data_store.py
Comment (hunterhector) [inline forte/data/data_store.py:527]: Should you be checking the `deletion_count` inside the for loop?

File: tests/forte/data/data_store_test.py
Comment (hunterhector) [inline tests/forte/data/data_store_test.py:349]: add a few more tests:
1. delete more than one entry
2. delete an entry of a child type and count the number of the parent type
3. try delete and then add, and then count the numbers

File: tests/forte/data/data_store_test.py
Comment (hunterhector) [inline tests/forte/data/data_store_test.py:639]: if we remove these test cases, do we have test cases for `add`?

Comment (feipenghe) [reply tests/forte/data/data_store_test.py]: They are added from other PRs.

File: tests/forte/data/data_store_test.py
Comment (hunterhector) [reply tests/forte/data/data_store_test.py]: I think the rest of the PR is fine. What's the current status of this comment?

@mylibrar  feel free to resolve the comments if you are fine with them, and you can approve it after your review too.

**Review by hunterhector — COMMENTED** [review]
Looks good to me, but will leave it up to @mylibrar to resolve all comments and approve it.

**Review by mylibrar — APPROVED** [review]

**Pull request merged**
